main.cpp

```diff
@@ -1,75 +1,96 @@
-
-class CastCallBack : public MatchFinder::MatchCallback {
-public:
-    CastCallBack(Rewriter& rewriter) {
-        // Your code goes here
-    };
-
-    void run(const MatchFinder::MatchResult &Result) override {
-        // Your code goes here
-    }
-};
-
-class MyASTConsumer : public ASTConsumer {
-public:
-    MyASTConsumer(Rewriter &rewriter) : callback_(rewriter) {
-        matcher_.addMatcher(
-                cStyleCastExpr(unless(isExpansionInSystemHeader())).bind("cast"), &callback_);
-    }
-
-    void HandleTranslationUnit(ASTContext &Context) override {
-        matcher_.matchAST(Context);
-    }
-
-private:
-    CastCallBack callback_;
-    MatchFinder matcher_;
-};
-
-class CStyleCheckerFrontendAction : public ASTFrontendAction {
-public:
-    CStyleCheckerFrontendAction() = default;
-    
-    void EndSourceFileAction() override {
-        rewriter_.getEditBuffer(rewriter_.getSourceMgr().getMainFileID())
-            .write(llvm::outs());
-    }
-
-    std::unique_ptr<ASTConsumer> CreateASTConsumer(CompilerInstance &CI, StringRef /* file */) override {
-        rewriter_.setSourceMgr(CI.getSourceManager(), CI.getLangOpts());
-        return std::make_unique<MyASTConsumer>(rewriter_);
-    }
-
-private:
-    Rewriter rewriter_;
-};
-
-static llvm::cl::OptionCategory CastMatcherCategory("cast-matcher options");
-
-int main(int argc, const char **argv) {
-    // Ubuntu 20
-    auto Parser = llvm::ExitOnError()(CommonOptionsParser::create(argc, argv, CastMatcherCategory));
-    // For Ubuntu 18, use the old Clang API to construct the CommonOptionsParser:
-    // CommonOptionsParser Parser(argc, argv, CastMatcherCategory);
-
-    ClangTool Tool(Parser.getCompilations(), Parser.getSourcePathList());
-    return Tool.run(newFrontendActionFactory<CStyleCheckerFrontendAction>().get());
-}
+#include <iostream>
+#include <sstream>
+
+#include <llvm/Support/CommandLine.h>
+#include <llvm/Support/Error.h>
+
+#include <clang/ASTMatchers/ASTMatchFinder.h>
+#include <clang/ASTMatchers/ASTMatchers.h>
+#include <clang/Basic/Diagnostic.h>
+#include <clang/Frontend/CompilerInstance.h>
+#include <clang/Frontend/FrontendActions.h>
+#include <clang/Rewrite/Core/Rewriter.h>
+#include <clang/Tooling/CommonOptionsParser.h>
+#include <clang/Tooling/Tooling.h>
+
+using namespace clang;
+using namespace clang::ast_matchers;
+using namespace clang::tooling;
+
+class CastCallBack : public MatchFinder::MatchCallback {
+
+private:
+  Rewriter& _rewriter;
+
+public:
+  CastCallBack(Rewriter &rewriter) : _rewriter(rewriter){};
+  void run(const MatchFinder::MatchResult &Result) override {
+
+    const auto *castExpr = Result.Nodes.getNodeAs<CStyleCastExpr>("cast");
+    
+    const auto range = CharSourceRange::getCharRange(
+        castExpr->getLParenLoc(),
+        castExpr->getSubExprAsWritten()->getBeginLoc());
+
+    const auto &sourceManager = *Result.SourceManager;
+
+    const auto type = Lexer::getSourceText(
+        CharSourceRange::getTokenRange(
```

Comment:
Please descrbe the difference between the use of `getCharRange()` and `getTokenRange()` with code examples.

Reply:
getCharRange() returns the range of characters up to and including the last character.
getTokenRange() returns the range of tokens before the start of the last token.

To see the difference more clearly we can run something like:

```cpp
  auto charRange = Lexer::getAsCharRange(
      CharSourceRange::getCharRange(node->getLParenLoc().getLocWithOffset(1),
      node->getRParenLoc().getLocWithOffset(-2)),
      SM, Result.Context->getLangOpts());
  
  auto tokenRange = Lexer::getAsCharRange(
      CharSourceRange::getTokenRange(node->getLParenLoc().getLocWithOffset(1),
      node->getRParenLoc().getLocWithOffset(-2)),
      SM, Result.Context->getLangOpts());
  
  llvm::outs() << "\n char range start: " << charRange.getBegin().printToString(SM)
                     << ",\n char range end: " << charRange.getEnd().printToString(SM) << "\n";

  llvm::outs() << "\n token range start: " << tokenRange.getBegin().printToString(SM)
                     << ",\n token range end: " << tokenRange.getEnd().printToString(SM) << "\n";
```

on this code:

```cpp
 int main() {
     int i = (int)     1.234;
     return 0;
 }
```

The result will be:

```
   char range start: /UNN-C-Style-Checker/main.cpp:3:14,
   char range end: /UNN-C-Style-Checker/main.cpp:3:15
  
   token range start: /UNN-C-Style-Checker/main.cpp:3:14,
   token range end: /UNN-C-Style-Checker/main.cpp:3:17
```

Reply:
• `getCharRange()` returns range specifying the starting/ending character by the specified locations.
  • `getTokenRange()`  returns range specifying the start of the range and the start of the last token of the range.